source/reduce/operand_to_dominating_id_reduction_pass.cpp

```diff
@@ -0,0 +1,118 @@
+
+using namespace opt;
+
+std::vector<std::unique_ptr<ReductionOpportunity>>
+OperandToDominatingIdReductionPass::GetAvailableOpportunities(
+    opt::IRContext* context) const {
+  std::vector<std::unique_ptr<ReductionOpportunity>> result;
+
+  // Go through every instruction in every block, considering it as a potential
+  // dominator of other instructions.  We choose this order for two reasons:
+  //
+  // (1) it is profitable for multiple opportunities to replace the same id x by
+  // different dominating ids y and z to be discontiguous, as they are
+  // incompatible.
+  //
+  // (2) We want to prioritise opportunities to replace an id with a more
+  // distant dominator.  Intuitively, in a human-readable programming language
+  // if we have a complex expression e with many sub-expressions, we would like
+  // to prioritise replacing e with its smallest sub-expressions; generalising
+  // this idea to dominating ids this roughly corresponds to more distant
+  // dominators.
+  for (auto& function : *context->module()) {
+    for (auto dominating_block = function.begin();
+         dominating_block != function.end(); ++dominating_block) {
+      for (auto& dominating_inst : *dominating_block) {
+        if (dominating_inst.HasResultId() && dominating_inst.type_id()) {
+          // Consider replacing any operand with matching type in a dominated
+          // instruction with the id generated by this instruction.
+          GetOpportunitiesForDominatingInst(
+              &result, &dominating_inst, dominating_block, &function, context);
+        }
+      }
+    }
+  }
+  return result;
+}
+
+void OperandToDominatingIdReductionPass::GetOpportunitiesForDominatingInst(
+    std::vector<std::unique_ptr<ReductionOpportunity>>* opportunities,
+    opt::Instruction* candidate_dominator,
+    opt::Function::iterator candidate_dominator_block, opt::Function* function,
+    opt::IRContext* context) const {
+  assert(candidate_dominator->HasResultId());
+  assert(candidate_dominator->type_id());
+  auto dominator_analysis = context->GetDominatorAnalysis(function);
+  // SPIR-V requires a block to precede all blocks it dominates, so it suffices
+  // to search from the candidate dominator block onwards.
+  for (auto block = candidate_dominator_block; block != function->end();
+       ++block) {
+    if (!dominator_analysis->Dominates(&*candidate_dominator_block, &*block)) {
+      // If the candidate dominator block doesn't dominate this block then there
+      // cannot be any of the desired reduction opportunities in this block.
+      continue;
+    }
+    for (auto& inst : *block) {
+      // We iterate through the operands using an explicit index (rather
+      // than using a lambda) so that we use said index in the construction
+      // of a ChangeOperandReductionOpportunity
+      for (uint32_t index = 0; index < inst.NumOperands(); index++) {
+        const auto& operand = inst.GetOperand(index);
+        if (spvIsIdType(operand.type)) {
```

Comment:
This is starting to look like a useful helper function for you. I would suggest spvIsInIdType(). Please update WhileEachInId in addition to the other reduction passes.

Reply:
Doing a helper sounds good.  What's the significance of "In" in the name, though, in terms of distinguishing the function from the existing spvIsIdType()?

Reply:
The ids it blacklists are related to the result of an instruction so this function returns true for any input operand id.

Reply:
OK, makes sense!